### ISSUE TITLE
self.__class__.__name__ has type Any in class derived from Any

### ISSUE THREAD

**Issue by gvanrossum** [python/mypy#4058]
Here's something that came up in https://github.com/python/typeshed/pull/1549#issuecomment-333227407.
```py
from typing import Any, Type, TypeVar
T = TypeVar('T')
class C(Any):
    def bar(self: T) -> Type[T]: ...
    def foo(self) -> None:
        reveal_type(self.bar())  # Type[C], as expected
        reveal_type(self.bar().__name__)  # Any, but expected str
```
@ilevkivskyi I think you muttered something about `analyze_member_access` for `TypeType` still being broken but I'm not sure I understand.

**Comment by ilevkivskyi** [comment]
This will be fixed by #4038

**Comment by gvanrossum** [comment]
Ah, thanks. I'm waiting for Jukka to review that, it's a bit above my review capacity ATM.

**Comment by ilevkivskyi** [comment]
I will add this example to the tests in the meantime.

**Comment by gvanrossum** [comment]
FWIW, the issue in our internal codebase that prompted this report is not resolved by #4038. But that issue is more complex than my example here, and I haven't boiled it down yet. To avoid referencing the other (closed, typeshed) issue, here is the problem again.

Some file gives this error:
```
[...]/task.py:221: error: Cannot determine type of '__name__'
```
The source code at that line looks like this:
```py
from threading import local
...
class TrackedTask(local):
    def __init__(self, ...):
        # type: ...
        if SupportedTrackedTaskWorkers.is_supported(self.__class__.__name__, self.task_name): ...
```

**Comment by ilevkivskyi** [comment]
Hm, strange, I cant reproduce this. I replaced ``SupportedTrackedTaskWorkers.is_supported`` with ``len`` and removed the second argument and everything works, also ``reveal_type(self.__class__.__name__)`` shows ``builtns.str``. I tried other combinations but still can't get this error.

**Comment by gvanrossum** [comment]
Me neither. :-( There must be something else in the file that affects this, but I don't yet know what it is. There must be something really strange going on, since if I add `reveal_type(self.__class__.__name__)` I get both these:
```
devtools/changes/changes/queue/task.py:221: error: Revealed type is 'builtins.str'
devtools/changes/changes/queue/task.py:221: error: Cannot determine type of '__name__'
```

**Comment by JelleZijlstra** [comment]
Regardless of the possible mypy bug, could we fix this in typeshed by making threading.local not inherit from Any and instead override `__setattr__` and `__getattr__`?

**Comment by gvanrossum** [comment]
Yes, that should be fine now.

**Comment by gvanrossum** [comment]
Wow. Thanks!
